File: cerebras/test/test_minimum.py
import torch, pytest
import custom_module

@pytest.mark.parametrize("input, other", [
    (torch.tensor([3.0, 0.0, -2.0]), torch.tensor([3.0, 0.0, 3.0])),   
    (torch.tensor([5.0]), torch.tensor([-3.0])), 
    (torch.tensor([1.5, -1.5]), torch.tensor([3.0, 2.0])),  
    (torch.tensor([0.0]), torch.tensor([0.0])),  
])
def test_minimum(input, other):

    # Create an instance of Minimum
    ch = custom_module.Minimum()

    # Perform custom Minimum
    custom_minimum = ch.compute_min(input, other)

    torch_minimum = torch.minimum(input, other)


    # Assert that the custom result matches the expected output
    assert torch.allclose(custom_minimum, torch_minimum, rtol=1e-05, atol=1e-05), \
    f"custom_minimum({input}, {other}) != torch.minimum({input}, {other})"

File: cerebras/gtest/test_minimum.cpp
// test_minimum.cpp
#include <gtest/gtest.h>
#include "minimum.h"  // Include the custom Minimum class header

// Test case for tensors with matching shapes and element-wise minimum
TEST(MinimumTest, MatchingShape) {
    torch::Tensor tensor1 = torch::tensor({1.0, 5.0, 3.0});
    torch::Tensor tensor2 = torch::tensor({4.0, 2.0, 6.0});
    torch::Tensor expected = torch::minimum(tensor1, tensor2);

    custom_namespace::Minimum min_instance;  // Create an instance of Minimum
    auto result = min_instance.compute_min(tensor1, tensor2);
    EXPECT_TRUE(torch::allclose(result, expected));  // Verifies that result matches expected tensor
}

// The rest of the tests can follow the same pattern:
TEST(MinimumTest, MismatchedShape) {
    torch::Tensor tensor1 = torch::tensor({1.0, 5.0, 3.0});
    torch::Tensor tensor2 = torch::tensor({4.0, 2.0});

    custom_namespace::Minimum min_instance;
    EXPECT_THROW(min_instance.compute_min(tensor1, tensor2), std::invalid_argument);
}

TEST(MinimumTest, NegativeValues) {
    torch::Tensor tensor1 = torch::tensor({-1.0, -5.0, 3.0});
    torch::Tensor tensor2 = torch::tensor({-4.0, 2.0, -6.0});
    torch::Tensor expected = torch::minimum(tensor1, tensor2);
  

    custom_namespace::Minimum min_instance;
    auto result = min_instance.compute_min(tensor1, tensor2);
    EXPECT_TRUE(torch::allclose(result, expected, 1e-3));
}

TEST(MinimumTest, EqualTensors) {
    torch::Tensor tensor1 = torch::tensor({3.0, 5.0, 1.0});
    torch::Tensor tensor2 = torch::tensor({3.0, 5.0, 1.0});
    torch::Tensor expected = torch::minimum(tensor1, tensor2);

    custom_namespace::Minimum min_instance;
    auto result = min_instance.compute_min(tensor1, tensor2);
    EXPECT_TRUE(torch::allclose(result, expected));
}
